build-support/bin/packages.py

```diff
@@ -74,12 +74,95 @@ def owners(self) -> set[str]:
         return {row[1] for row in roles if row[0] == "Owner"}  # type: ignore[union-attr,index]
 
 
+def _pip_args(extra_pip_args: list[str]) -> tuple[str, ...]:
+    return (*extra_pip_args, "--quiet", "--no-cache-dir")
+
+
+def validate_pants_pkg(version: str, venv_dir: Path, extra_pip_args: list[str]) -> None:
+    subprocess.run(
+        [
+            venv_dir / "bin/pip",
+            "install",
+            *_pip_args(extra_pip_args),
+            f"pantsbuild.pants=={version}",
+        ],
+        check=True,
+    )
+
+    def run_venv_pants(args: list[str]) -> str:
+        # When we do (dry-run) testing, we need to run the packaged pants. It doesn't have internal
+        # backend plugins so when we execute it at the repo build root, the root pants.toml will
+        # ask it to load internal backend packages and their dependencies which it doesn't have,
+        # and it'll fail. To solve that problem, we load the internal backend package dependencies
+        # into the pantsbuild.pants venv.
+        return (
+            subprocess.run(
+                [
+                    venv_dir / "bin/pants",
+                    "--no-verify-config",
+                    "--no-remote-cache-read",
+                    "--no-remote-cache-write",
+                    "--no-pantsd",
+                    "--pythonpath=['pants-plugins']",
+                    (
+                        "--backend-packages=["
+                        "'pants.backend.awslambda.python', "
+                        "'pants.backend.python', "
+                        "'pants.backend.shell', "
+                        "'internal_plugins.releases'"
+                        "]"
+                    ),
+                    *args,
+                ],
+                check=True,
+                stdout=subprocess.PIPE,
```

Comment:
Doesn't this call also need `capture_output=True`? https://docs.python.org/3/library/subprocess.html#subprocess.run

Reply:
No, that kwarg is a shorthand. I don't want to capture stderr because we want errors to be printed, and this current factoring avoids doing that.

> If capture_output is true, stdout and stderr will be captured. When used, the internal Popen object is automatically created with stdout=PIPE and stderr=PIPE. The stdout and stderr arguments may not be supplied at the same time as capture_output. If you wish to capture and combine both streams into one, use stdout=PIPE and stderr=STDOUT instead of capture_output.